File: src/ru/edu/vtisov/string/EncodeAndDecodeStrings.java
package ru.edu.vtisov.string;

// https://leetcode.com/problems/encode-and-decode-strings

import java.util.ArrayList;
import java.util.List;

public class EncodeAndDecodeStrings {

    private static final String DELIMETER = "/:";

    // Encodes a list of strings to a single string.
    public String encode(List<String> strs) {
        StringBuilder encoded = new StringBuilder();
        for (String str: strs) {
            encoded.append(str.length() + DELIMETER + str);
        }
        return encoded.toString();
    }

    // Decodes a single string to a list of strings.
    public List<String> decode(String s) {
        List<String> decoded = new ArrayList<>();
        int i = 0;
        while (i < s.length()) {
            int delimeter = s.indexOf(DELIMETER, i);
            int length = Integer.parseInt(s.substring(i, delimeter));
            decoded.add(s.substring(delimeter + 2, delimeter + 2 + length));
            i = delimeter + 2 + length;
        }
        return decoded;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.decode(codec.encode(strs));

File: src/ru/edu/vtisov/array/encode-and-decode-strings.go
package array

// https://leetcode.com/problems/encode-and-decode-strings

import "strings"

type Codec struct{}

func (Codec) Encode(strs []string) string {
	var buffer = strings.Builder{}
	for _, str := range strs {
		buffer.WriteString(strings.ReplaceAll(str, "/", "//"))
		buffer.WriteString("/:")
	}
	return buffer.String()
}

func (Codec) Decode(s string) []string {
	var result = []string{}
	var buffer = strings.Builder{}

	for i := 0; i < len(s); {
		if i+1 < len(s) && s[i] == '/' && s[i+1] == '/' {
			buffer.WriteByte('/')
			i += 2
		} else if i+1 < len(s) && s[i] == '/' && s[i+1] == ':' {
			result = append(result, buffer.String())
			buffer.Reset()
			i += 2
		} else {
			buffer.WriteByte(s[i])
			i++
		}
	}
	return result
}
